### ISSUE TITLE
Larger loss on GPU when using PLaidML vs without on CPU

### ISSUE THREAD

**Issue by CharcoalStyles** [plaidml/plaidml#173]
I'm playing around with Variational Autoencoders and have noticed that when running it on GPUs with PlaidML it doesn't learn as well as when I run it on CPU.

Here are outputs from my work computer:

With PlaidML (on GTX 750m)
`Epoch 1/2
60000/60000 [==============================] - 32s 532us/step - loss: 224.1040 - val_loss: 206.4646
Epoch 2/2
60000/60000 [==============================] - 14s 241us/step - loss: 206.6054 - val_loss: 206.2648`
[Full Log](https://pastebin.com/Tj0UxSEU)

Without PlaidML (on Core i7-4712HQ)
`Epoch 1/2
60000/60000 [==============================] - 25s 409us/step - loss: 179.9514 - val_loss: 157.2696
Epoch 2/2
60000/60000 [==============================] - 28s 472us/step - loss: 154.0158 - val_loss: 151.6601`
[Full Log](https://pastebin.com/Tj0UxSEU)

Running more epochs shows a similar difference in loss, with still no convergence after ~100.

I can replicate at home on my desktop with it's i5-2550K and Radeon R9-380 (which is the reason I found PLaidML in the first place).

**Comment by colorfuldarkgray** [comment]
I also noticed that using Gpu with plaidml is slower than cpu only. Could somebody tell in what cases plaidml is worthy to use?

inceptionv3 using plaidml ("finally" using radeon gpu):
{"compile_duration": 82.23041248321533, "execution_duration": 1.793085461948067, "example_size": 1024, "batch_size": 1, "model": "inception_v3", "correct": true, "max_error": 1.598830749571789e-05, "max_abs_error": 1.773238182067871e-06, "fail_ratio": 0.0}

inceptionv3 using tf (cpu only):
{"compile_duration": 5.7633442878723145, "execution_duration": 0.9607016900554299, "example_size": 1024, "batch_size": 1, "model": "inception_v3", "correct": true, "max_error": 1.008163144433638e-05, "max_abs_error": 5.21540641784668e-07, "fail_ratio": 0.0}

**Comment by brianretford** [comment]
@colorfuldarkgray Can you provide more details on your system configuration? Modern CPUs are quite competitive with many GPUs at batch_size 1, if you bump the batch size up you should hopefully see a more positive difference.

**Comment by CharcoalStyles** [comment]
@colorfuldarkgray The model I was training was based off [this sample](https://github.com/keras-team/keras/blob/master/examples/variational_autoencoder.py), with the same batch size of 128.

The layers are in the full logs, but here are the plots.

- [Encoder](https://imgur.com/hBiZcPi)
- [Decoder](https://imgur.com/AfixUPj)
- [VAE](https://imgur.com/KqqTG7O)

As you can see, it's fairly simple, just a few Dense layers.

I'll do the normal MNIST training now and post up the results.

**Comment by CharcoalStyles** [comment]
@colorfuldarkgray I did the MNIST classification; dual dense layers, with dropout. [Link to model plot](https://imgur.com/a/DkJSfM0)

With PlaidML (on GTX 750m)
`Epoch 1/25
60000/60000 [==============================] - 7s 115us/step - loss: 0.3834 - acc: 0.8882 - val_loss: 0.1451 - val_acc: 0.9552
Epoch 2/25
60000/60000 [==============================] - 4s 61us/step - loss: 0.1407 - acc: 0.9582 - val_loss: 0.0984 - val_acc: 0.9693
...
Epoch 24/25
60000/60000 [==============================] - 4s 61us/step - loss: 0.0113 - acc: 0.9960 - val_loss: 0.0676 - val_acc: 0.9832
Epoch 25/25
60000/60000 [==============================] - 4s 60us/step - loss: 0.0101 - acc: 0.9967 - val_loss: 0.0691 - val_acc: 0.9846`
[Full Log](https://pastebin.com/hnXpJt6P)

Without PlaidML (on Core i7-4712HQ)
`Epoch 1/25
60000/60000 [==============================] - 5s 89us/step - loss: 0.3960 - acc: 0.8833 - val_loss: 0.1464 - val_acc: 0.9571
Epoch 2/25
60000/60000 [==============================] - 5s 75us/step - loss: 0.1420 - acc: 0.9571 - val_loss: 0.1009 - val_acc: 0.9683
...
Epoch 24/25
60000/60000 [==============================] - 5s 77us/step - loss: 0.0131 - acc: 0.9954 - val_loss: 0.0696 - val_acc: 0.9830
Epoch 25/25
60000/60000 [==============================] - 5s 78us/step - loss: 0.0101 - acc: 0.9963 - val_loss: 0.0736 - val_acc: 0.9819`
[Full Log](https://pastebin.com/Udh8ECdH)

This looks all ok, if you ignore the massive over-fitting ;)

So I'm guessing it's something with the VAE & PlaidML.

**Comment by colorfuldarkgray** [comment]
@brianretford @CharcoalStyles I'm sorry, I realize I made a rude comment. Whats is not worthy is trying to use some AMD 6310 with R4 gpu to run NN. I can't try to run other nets right now. The results were given by plaidbench. I ran that after I felt plaidml was not helping much when playing a bit with keras and inception.
